src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/reports/ReconciliationService.java

```diff
@@ -0,0 +1,114 @@
+package uk.gov.hmcts.reform.bulkscanprocessor.services.reports;
+
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.bulkscanprocessor.entity.reports.ReceivedZipFile;
+import uk.gov.hmcts.reform.bulkscanprocessor.entity.reports.ReceivedZipFileRepository;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.Discrepancy;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.ReconciliationStatement;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.ZipFileData;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.stream.Collectors.toMap;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.PAYMENT_DCNS_MISMATCH;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.RECEIVED_BUT_NOT_REPORTED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.REPORTED_BUT_NOT_RECEIVED;
+import static uk.gov.hmcts.reform.bulkscanprocessor.services.reports.models.DiscrepancyType.SCANNABLE_DOCUMENT_DCNS_MISMATCH;
+
+@Service
+public class ReconciliationService {
+    private final ReceivedZipFileRepository receivedZipFileRepository;
+    private final ReceivedZipFileConverter receivedZipFileConverter;
+
+    public ReconciliationService(
+        ReceivedZipFileRepository receivedZipFileRepository,
+        ReceivedZipFileConverter receivedZipFileConverter
+    ) {
+        this.receivedZipFileRepository = receivedZipFileRepository;
+        this.receivedZipFileConverter = receivedZipFileConverter;
+    }
+
+    public List<Discrepancy> getReconciliationReport(ReconciliationStatement reconciliationStatement) {
+        List<ReceivedZipFile> receivedZipFiles =
+            receivedZipFileRepository.getReceivedZipFilesReportFor(reconciliationStatement.date);
+        List<ZipFileData> receivedZipFileDataList =
+            new ArrayList<>(receivedZipFileConverter.convertReceivedZipFiles(receivedZipFiles));
+
+        List<Discrepancy> discrepancies = new ArrayList<>();
+
+        Map<ZipFileData, ZipFileData> envelopesMap =
+            reconciliationStatement.envelopes.stream().collect(toMap(envelope -> envelope, envelope -> envelope));
```

Comment:
what is this map for?

Reply:
This is to provide access to the reported envelopes. If we use list we can find (`contains(...)`) if the object is in the list (but inefficiently performance wise) but cannot get access to this object. Using map we can find if the object is in the map (`containsKey(...)`easily and easily get access to it. Keep in mind that `ZipFileData` class is essential here - it provides (thank to its `equals` and `hashCode` methods) comparison of reported and received file (`ReportedZipFileData` and `ReceivedZipFile`).

Reply:
why not a set?

Reply:
Because there is no `get` operation in `Set`. When `Set.contains(...)` is called in this case 2 different objects are compared - reported file and received file, and the comparison operation compares `zipFileName` and `container`. But if `contains` returns `true` (i.e. reported files have an entry with the same `zipFileName` and `container` as received files) we need to extract (i.e. `get`) this reported file to compare fields (`scannableDoumentDcns` and `paymentDcns`). `Map` has this `get` operation, `Set` does not.